src/cache-over-fs.ts

```diff
@@ -0,0 +1,74 @@
+import * as Promise from 'bluebird';
+import {
+    FileSystem,
+    Directory,
+    fileSystemMethods
+} from "./api";
+
+import {MemoryFileSystem} from "./memory-fs";
+import {InternalEventsEmitter, makeEventsEmitter} from "./utils";
+
+type CalledMethods = {
+    [prop: string]: boolean
+};
+
+export class CacheOverFs implements FileSystem {
+    public readonly events: InternalEventsEmitter;
+    public baseUrl: string;
+    private cache: FileSystem;
+    public readonly fs: FileSystem;
+    private calledMethods: CalledMethods;
+
+    constructor(fs: FileSystem) {
+        this.fs = fs;
```

Comment:
no need to explicitly define field this.fs
https://www.stevefenton.co.uk/2013/04/stop-manually-assigning-typescript-constructor-parameters/